mods/sfinv/api.lua

```diff
@@ -0,0 +1,158 @@
+}
+
+function sfinv.register_page(name, def)
+	assert(name, "Invalid sfinv page. Requires a name")
+	assert(def, "Invalid sfinv page. Requires a def[inition] table")
+	assert(def.get, "Invalid sfinv page. Def requires a get function.")
+	assert(not sfinv.pages[name], "Attempt to register already registered sfinv page " .. dump(name))
+
+	sfinv.pages[name] = def
+	def.name = name
+	table.insert(sfinv.pages_unordered, def)
+end
+
+function sfinv.override_page(name, def)
+	assert(name, "Invalid sfinv page override. Requires a name")
+	assert(def, "Invalid sfinv page override. Requires a def[inition] table")
+	local page = sfinv.pages[name]
+	assert(page, "Attempt to override sfinv page " .. dump(name) .. " which does not exist.")
+	for key, value in pairs(def) do
+		page[key] = value
+	end
+end
+
+function sfinv.get_nav_fs(player, context, nav, current_idx)
+	-- Only show tabs if there is more than one page
+	if #nav > 1 then
+		return "tabheader[0,0;tabs;" .. table.concat(nav, ",") .. ";" .. current_idx .. ";true;false]"
+	else
+		return ""
+	end
+end
+
+local theme_main = "bgcolor[#080808BB;true]" .. default.gui_bg ..
+		default.gui_bg_img
+
+local theme_inv = default.gui_slots .. [[
+		list[current_player;main;0,4.7;8,1;]
+		list[current_player;main;0,5.85;8,3;8]
+	]]
+
+function sfinv.make_formspec(player, context, content, show_inv, size)
+	local tmp = {
+		size or "size[8,8.6]",
+		theme_main,
+		sfinv.get_nav_fs(player, context, context.nav_titles, context.nav_idx),
+		content
+	}
+	if show_inv then
+		tmp[#tmp + 1] = theme_inv
+	end
+	return table.concat(tmp, "")
+end
+
+function sfinv.get_formspec(player, context)
+	-- Generate navigation tabs
+	local nav = {}
+	local nav_ids = {}
+	local current_idx = 1
+	for i, pdef in pairs(sfinv.pages_unordered) do
+		if not pdef.is_in_nav or pdef:is_in_nav(player, context) then
+			nav[#nav + 1] = pdef.title
+			nav_ids[#nav_ids + 1] = pdef.name
+			if pdef.name == context.page then
+				current_idx = i
+			end
+		end
+	end
+	context.nav = nav_ids
+	context.nav_titles = nav
+	context.nav_idx = current_idx
+
+	-- Generate formspec
+	local page = sfinv.pages[context.page] or sfinv.pages["404"]
+	if page then
+		return page:get(player, context)
+	else
+		local old_page = context.page
+		context.page = sfinv.homepage_name
+		assert(sfinv.pages[context.page], "[sfinv] Invalid homepage")
+		minetest.log("warning", "[sfinv] Couldn't find " .. dump(old_page) .. " so using switching to homepage")
+		return sfinv.get_formspec(player, context)
+	end
+end
+
+function sfinv.set(player, context)
+	if not context then
+		local name = player:get_player_name()
+		context = sfinv.contexts[name]
+		if not context then
+			context = {
+				page = sfinv.homepage_name
+			}
+			sfinv.contexts[name] = context
+		end
+	end
+
+	local fs = sfinv.get_formspec(player, context)
+	player:set_inventory_formspec(fs)
+end
+
+minetest.register_on_joinplayer(function(player)
+	if sfinv.enabled then
+		minetest.after(0.5, function()
```

Comment:
What is the `minetest.after` for?

The delayed initialization leads to the creative inventory to override any custom inventories on every join, including unified inventory.

If that is intentional, you should probably mention in the PR text, how inventory mods have to be updated to work.

MTG will essentially remove `set_inventory_formspec` from the API by rendering it resultless, which i think is noteworthy in the documentation as well, and should be accompanied by replacing it with a function that throws a warning when used by other mods, so that issues caused by this can be resolved swiftly.

Alternatively you could replace it with a function that will deactivate sfinv when being used. (sfinv itself can always keep the original function for its own use and a reset) That way you could even remain backward compatible.